File: src/main/java/com/github/zyypj/tadeuBooter/api/reflection/ReflectionHelper.java
package com.github.zyypj.tadeuBooter.api.reflection;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class ReflectionHelper {

    /**
     * Obtém o valor de um campo (privado ou público) de um objeto.
     */
    public static Object getFieldValue(Object target, String fieldName) throws NoSuchFieldException, IllegalAccessException {
        Field field = getDeclaredField(target.getClass(), fieldName);
        field.setAccessible(true);
        return field.get(target);
    }

    /**
     * Define o valor de um campo (privado ou público) de um objeto.
     */
    public static void setFieldValue(Object target, String fieldName, Object value) throws NoSuchFieldException, IllegalAccessException {
        Field field = getDeclaredField(target.getClass(), fieldName);
        field.setAccessible(true);
        field.set(target, value);
    }

    /**
     * Invoca um método (privado ou público) de um objeto.
     */
    public static Object invokeMethod(Object target, String methodName, Class<?>[] parameterTypes, Object... args) throws Exception {
        Method method = getDeclaredMethod(target.getClass(), methodName, parameterTypes);
        method.setAccessible(true);
        return method.invoke(target, args);
    }

    private static Field getDeclaredField(Class<?> clazz, String fieldName) throws NoSuchFieldException {
        Class<?> current = clazz;
        while (current != null) {
            try {
                return current.getDeclaredField(fieldName);
            } catch (NoSuchFieldException e) {
                current = current.getSuperclass();
            }
        }
        throw new NoSuchFieldException("Field " + fieldName + " not found in class " + clazz.getName());
    }

    private static Method getDeclaredMethod(Class<?> clazz, String methodName, Class<?>[] parameterTypes) throws NoSuchMethodException {
        Class<?> current = clazz;
        while (current != null) {
            try {
                return current.getDeclaredMethod(methodName, parameterTypes);
            } catch (NoSuchMethodException e) {
                current = current.getSuperclass();
            }
        }
        throw new NoSuchMethodException("Method " + methodName + " not found in class " + clazz.getName());
    }
}

File: shared/src/main/java/me/zyypj/booter/shared/reflection/ReflectionHelper.kt
package me.zyypj.booter.shared.reflection

import java.lang.reflect.Field
import java.lang.reflect.Method

object ReflectionHelper {

    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    fun getFieldValue(target: Any, fieldName: String): Any? {
        val field = getDeclaredField(target.javaClass, fieldName)
        field.isAccessible = true
        return field.get(target)
    }

    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    fun setFieldValue(target: Any, fieldName: String, value: Any?) {
        val field = getDeclaredField(target.javaClass, fieldName)
        field.isAccessible = true
        field.set(target, value)
    }

    @Throws(Exception::class)
    fun invokeMethod(
        target: Any,
        methodName: String,
        parameterTypes: Array<Class<*>>,
        vararg args: Any?
    ): Any? {
        val method = getDeclaredMethod(target.javaClass, methodName, parameterTypes)
        method.isAccessible = true
        return method.invoke(target, *args)
    }

    @Throws(NoSuchFieldException::class)
    private fun getDeclaredField(clazz: Class<*>, fieldName: String): Field {
        var current: Class<*>? = clazz
        while (current != null) {
            try {
                return current.getDeclaredField(fieldName)
            } catch (_: NoSuchFieldException) {
                current = current.superclass
            }
        }
        throw NoSuchFieldException("Field $fieldName não encontrado na classe ${clazz.name}")
    }

    @Throws(NoSuchMethodException::class)
    private fun getDeclaredMethod(
        clazz: Class<*>,
        methodName: String,
        parameterTypes: Array<Class<*>>
    ): Method {
        var current: Class<*>? = clazz
        while (current != null) {
            try {
                return current.getDeclaredMethod(methodName, *parameterTypes)
            } catch (_: NoSuchMethodException) {
                current = current.superclass
            }
        }
        throw NoSuchMethodException("Method $methodName não encontrado na classe ${clazz.name}")
    }
}